src/main/java/io/numaproj/numaflow/function/FunctionService.java

```diff
@@ -9,44 +14,37 @@
 import io.numaproj.numaflow.function.metadata.IntervalWindowImpl;
 import io.numaproj.numaflow.function.metadata.Metadata;
 import io.numaproj.numaflow.function.metadata.MetadataImpl;
-import io.numaproj.numaflow.function.reduce.ReduceDatumStream;
-import io.numaproj.numaflow.function.reduce.ReduceDatumStreamImpl;
-import io.numaproj.numaflow.function.reduce.ReduceHandler;
+import io.numaproj.numaflow.function.reduce.GroupBy;
+import io.numaproj.numaflow.function.reduce.ReduceSupervisorActor;
+import io.numaproj.numaflow.function.reduce.ShutdownActor;
 import io.numaproj.numaflow.function.v1.Udfunction;
 import io.numaproj.numaflow.function.v1.Udfunction.EventTime;
 import io.numaproj.numaflow.function.v1.UserDefinedFunctionGrpc;
+import lombok.NoArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import scala.concurrent.Await;
+import scala.concurrent.Future;
+import scala.concurrent.duration.Duration;
 
 import java.time.Instant;
-import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
-import java.util.Map;
-import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.ExecutionException;
-import java.util.concurrent.ExecutorService;
-import java.util.concurrent.Executors;
-import java.util.concurrent.Future;
-import java.util.concurrent.TimeUnit;
-import java.util.logging.Level;
-import java.util.logging.Logger;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeoutException;
 
+import static io.numaproj.numaflow.function.Function.EOF;
 import static io.numaproj.numaflow.function.v1.UserDefinedFunctionGrpc.getMapFnMethod;
 import static io.numaproj.numaflow.function.v1.UserDefinedFunctionGrpc.getReduceFnMethod;
 
-class FunctionService extends UserDefinedFunctionGrpc.UserDefinedFunctionImplBase {
-    private static final Logger logger = Logger.getLogger(FunctionService.class.getName());
-    // it will never be smaller than one
-    private final ExecutorService reduceTaskExecutor = Executors
-            .newCachedThreadPool();
+@Slf4j
+@NoArgsConstructor
+public class FunctionService extends UserDefinedFunctionGrpc.UserDefinedFunctionImplBase {
```

Comment:
Do we have unit tests for this class? We might want to add one. Not necessarily within this PR.

Reply:
the FunctionServer.test covers the FunctionService part, but we might have to add some tests.